### PR TITLE
[LAA] Use PSE::getSymbolicMaxBackedgeTakenCount.

### PR DESCRIPTION
**Pull request by fhahn**
Update LAA to use PSE::getSymbolicMaxBackedgeTakenCount which returns
the minimum of the countable exits.

When analyzing dependences and computing runtime checks, we need the
smallest upper bound on the number of iterations. In terms of memory
safety, it shouldn't matter if any uncomputable exits leave the loop,
as long as we prove that there are no dependences given the minimum of
the countable exits. The same should apply also for generating runtime
checks.

Note that this shifts the responsiblity of checking whether all exit
counts are computable or handling early-exits to the users of LAA.

Depends on https://github.com/llvm/llvm-project/pull/93498

### CODE REVIEW

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 1]
```diff
@@ -2055,9 +2055,9 @@ MemoryDepChecker::Dependence::DepType MemoryDepChecker::isDependent(
   // stride multiplied by the backedge taken count, the accesses are independet,
   // i.e. they are far enough appart that accesses won't access the same
   // location across all loop ierations.
-  if (HasSameSize &&
-      isSafeDependenceDistance(DL, SE, *(PSE.getBackedgeTakenCount()), *Dist,
-                               MaxStride, TypeByteSize))
+  if (HasSameSize && isSafeDependenceDistance(
```
Comment (preames) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2058]: The doc comment on isSafeDependenceDistance needs updated.  I think it's correct, but there's a difference between an exact BTC and a bound on BTC.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 2]
```diff
@@ -2395,7 +2395,7 @@ bool LoopAccessInfo::canAnalyzeLoop() {
   }
 
   // ScalarEvolution needs to be able to find the exit count.
-  const SCEV *ExitCount = PSE->getBackedgeTakenCount();
+  const SCEV *ExitCount = PSE->getSymbolicMaxBackedgeTakenCount();
```
Comment (preames) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:2398]: Update the comments to say bound on the btc.

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp [hunk 3]
```diff
@@ -3004,7 +3004,7 @@ void LoopAccessInfo::collectStridedAccess(Value *MemAccess) {
   // of various possible stride specializations, considering the alternatives
   // of using gather/scatters (if available).
 
-  const SCEV *BETakenCount = PSE->getBackedgeTakenCount();
+  const SCEV *BETakenCount = PSE->getSymbolicMaxBackedgeTakenCount();
```
Comment (preames) [inline llvm/lib/Analysis/LoopAccessAnalysis.cpp:3007]: Not related to your change - but this whole block of code is just weird.  This is basically proving a more precise trip count, why is it in LAA at all?  Wouldn't simply early exiting on small BTC loops be sufficient?

File: llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp [hunk 4]
```diff
@@ -1506,6 +1506,16 @@ bool LoopVectorizationLegality::canVectorize(bool UseVPlanNativePath) {
       return false;
   }
 
+  if (isa<SCEVCouldNotCompute>(PSE.getBackedgeTakenCount())) {
```
Comment (preames) [inline llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:1509]: What about the other users of LAA in tree?  Have you audited them?  If not, can you add bailouts to ensure we're not breaking anything with this transition?

File: llvm/lib/Analysis/LoopAccessAnalysis.cpp (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Analysis/LoopAccessAnalysis.cpp]: Yeah, let me take a look into this separately.

File: llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp (hunk 4 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/LoopVectorizationLegality.cpp:1509]: I checked `LoopVersioningLICM`, `LoopDistribute.cpp` & `LoopLoadElimination.cpp`, they all check for a single exiting block, which should be sufficient I think. I also added tests for loop-distribute & loop-load-elim.

**Review by preames — APPROVED** [review]
LGTM

And thanks for doing this and all the other cleanup on LAA recently.  You're making this code much more robust and more importantly, easy to reason through.  Thanks!

**Comment by fhahn** [comment]
> LGTM
> 
> And thanks for doing this and all the other cleanup on LAA recently. You're making this code much more robust and more importantly, easy to reason through. Thanks!

Thanks, hopefully more to come in the near future!

**Pull request merged**